### PR TITLE
Pressure Laplacian Smoothness: topology-aware surface Cp regularization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Surface pressure distributions are physically smooth everywhere except at the stagnation point and separation region. The model has no explicit smoothness constraint — the existing DCT frequency loss (PR #2184, merged) penalizes high-frequency content but operates in sorted-node space with arbitrary LE/TE ordering issues. A **graph-Laplacian smoothness loss** operating on actual mesh adjacency is topology-aware and avoids the ordering ambiguity.

For each pair of adjacent surface nodes (i, j) on the same foil, penalize `(p_pred_i - p_pred_j)^2 / dist(i,j)^2` — the squared spatial gradient of predicted pressure. This enforces that surface Cp varies smoothly along the airfoil, which is a physically correct constraint everywhere except at stagnation and separation.

**Physical motivation:** Inviscid pressure distributions satisfy the Laplace equation (harmonic functions). Even in viscous flows at Re=1-5M, surface pressure is smooth between the stagnation point and separation. The mesh has ~150 surface nodes per airfoil — the Laplacian reduces to a finite difference on a 1D ring topology.

**Key distinction from failed loss modifications:**
- Bernoulli loss (#2224): imposed WRONG physics (Bernoulli invalid at viscous walls) → catastrophic
- Arc-length surface loss (#2210): conflicted with hard-node mining, used arbitrary ordering → p_in +14.2%
- DCT freq loss (#2184, merged): works in FREQUENCY domain with sorted-node ordering issues
- **This (Laplacian):** enforces TOPOLOGICAL smoothness in SPATIAL domain using actual mesh adjacency — physically correct, complementary to DCT

**Critical safeguard vs #2210 failure:** The arc-length loss failed because it interfered with the model's ability to produce sharp pressure gradients at LE and TE. This loss addresses that by **excluding high-curvature regions** (LE and TE vicinity) from the penalty. Nodes within 5% chord of LE or TE are exempt from the smoothness penalty, allowing sharp gradients where physics demands them.

**Confidence:** Medium. Physically motivated, topology-aware, complementary to DCT. Main risk: may be redundant with DCT freq loss if DCT already captures sufficient smoothness signal. The LE/TE exclusion zone addresses the #2210 failure mode.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

```python
pressure_laplacian_loss: bool = False    # topology-aware surface pressure smoothness
laplacian_weight: float = 0.01           # weight for Laplacian smoothness loss
laplacian_le_te_exclude: float = 0.05    # exclude nodes within this fraction of chord from LE/TE
```

### Step 2: Implement Laplacian smoothness computation

Add a helper function:

```python
def compute_pressure_laplacian_loss(p_pred, xy_coords, boundary_id, le_te_exclude=0.05):
    """Graph-Laplacian smoothness loss on surface pressure predictions.
    
    For each foil's surface nodes, sorted into a ring by angle from centroid,
    computes the squared pressure gradient between adjacent nodes.
    Excludes nodes near LE/TE (within le_te_exclude fraction of chord).
    
    Args:
        p_pred: [B, N] predicted pressure (or [B, N, 1])
        xy_coords: [B, N, 2] node coordinates
        boundary_id: [B, N] integer boundary IDs (6=fore surface, 7=aft surface)
        le_te_exclude: fraction of chord to exclude near LE and TE
    
    Returns: scalar loss (mean squared pressure gradient over all valid adjacent pairs)
    """
    if p_pred.dim() == 3:
        p_pred = p_pred.squeeze(-1)
    
    B = p_pred.shape[0]
    total_loss = torch.tensor(0.0, device=p_pred.device)
    count = 0
    
    for foil_bid in [6, 7]:
        for b in range(B):
            mask = (boundary_id[b] == foil_bid)
            if mask.sum() < 5:  # need enough nodes for meaningful smoothness
                continue
            
            xy = xy_coords[b, mask, :2]  # [M, 2]
            p = p_pred[b, mask]           # [M]
            
            # Sort by angle from centroid (ring ordering)
            centroid = xy.mean(dim=0, keepdim=True)
            angles = torch.atan2(xy[:, 1] - centroid[0, 1],
                                  xy[:, 0] - centroid[0, 0])
            order = angles.argsort()
            xy_sorted = xy[order]
            p_sorted = p[order]
            
            # Identify LE (min-x) and TE (max-x) positions
            x_min = xy[:, 0].min()
            x_max = xy[:, 0].max()
            chord = x_max - x_min + 1e-8
            
            # Create exclusion mask: exclude nodes near LE and TE
            x_sorted = xy_sorted[:, 0]
            not_le = (x_sorted - x_min) > le_te_exclude * chord
            not_te = (x_max - x_sorted) > le_te_exclude * chord
            valid = not_le & not_te  # [M]
            
            # Both node i and i+1 must be valid for the pair to count
            pair_valid = valid[:-1] & valid[1:]  # [M-1]
            
            if pair_valid.sum() < 3:
                continue
            
            # Squared pressure gradient between adjacent nodes
            dp = p_sorted[1:] - p_sorted[:-1]  # [M-1]
            ds = (xy_sorted[1:] - xy_sorted[:-1]).norm(dim=-1).clamp(min=1e-6)  # [M-1]
            grad_sq = (dp / ds) ** 2  # [M-1]
            
            # Only count valid pairs
            total_loss = total_loss + (grad_sq * pair_valid.float()).sum()
            count += pair_valid.sum().item()
    
    if count > 0:
        return total_loss / count
    return total_loss
```

### Step 3: Integrate into training loss

In the training step, after the main loss computation and before backward():

```python
if cfg.pressure_laplacian_loss:
    # Use the surface pressure predictions from the model
    # p_pred should be the predicted pressure for surface nodes
    lap_loss = compute_pressure_laplacian_loss(
        p_pred_surface, xy_coords, boundary_id, 
        le_te_exclude=cfg.laplacian_le_te_exclude
    )
    loss = loss + cfg.laplacian_weight * lap_loss
    
    # Log to W&B
    if wandb_log:
        log_dict['train/laplacian_loss'] = lap_loss.item()
```

**IMPORTANT:** Apply only in the training loop, not in validation/viz. The loss is a regularizer.

### Step 4: torch.compile compatibility

All operations (sort, indexing, norm, arithmetic) are compile-safe for fixed-shape tensors. The per-sample loop over `B` should be fine since B is small (typically 1-4). If torch.compile struggles with the variable `count`, move the normalization outside the compiled region.

### Step 5: Run 2 seeds with default weight

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/pressure-laplacian-s42" \
  --wandb_group "round18/pressure-laplacian-smooth" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --pressure_laplacian_loss --laplacian_weight 0.01

# Seed 73 — identical but --seed 73 --wandb_name "frieren/pressure-laplacian-s73"
```

### Step 6: Report results

Add a **Results** comment with:
- Table: p_in, p_oodc, p_tan, p_re for seed 42, seed 73, 2-seed avg
- Comparison against baseline
- W&B run IDs
- Log the `train/laplacian_loss` curve — note whether it decreases during training (model is learning to be smooth) or stays flat (already smooth, feature is redundant)
- If laplacian_loss is >5x the surface MAE loss magnitude, note this — may need to reduce weight

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: `hgml7i2r` (seed 42), `qic03vrg` (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Metrics (mae_surf_p)

| Metric | Baseline (PR #2213) | Seed 42 | Seed 73 | 2-seed Avg | Delta |
|--------|---------------------|---------|---------|------------|-------|
| **p_in** | **11.979** | 12.466 | 12.624 | 12.545 | **+4.7%** |
| **p_tan** | **28.341** | 55.121 | 55.936 | 55.529 | **+95.9%** |
| **p_oodc** | **7.643** | 31.837 | 30.351 | 31.094 | **+306.8%** |
| **p_re** | **6.300** | 24.822 | 24.479 | 24.651 | **+291.3%** |

### W&B Run IDs
- Seed 42: `85qhnb0q` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/85qhnb0q)
- Seed 73: `mtc0qkkc` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/mtc0qkkc)

### Command used
```bash
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/pressure-laplacian-s42" \
  --wandb_group "round18/pressure-laplacian-smooth" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --pressure_laplacian_loss --laplacian_weight 0.01
```
(Seed 73 identical with `--seed 73 --wandb_name "frieren/pressure-laplacian-s73"`)

Peak memory: ~60GB (GPU 0), ~76GB (GPU 1)

### Laplacian loss behavior
Final `train/laplacian_loss`: 6.2 (s42), 3.5 (s73). The loss did decrease during training (from ~14 down to ~3-6), indicating the model was learning to produce smoother pressure distributions. However, this smoothness constraint **catastrophically damaged OOD generalization**.

### What happened

**Catastrophic negative result.** The Laplacian smoothness loss destroyed performance on all OOD splits while barely affecting in-distribution:

- p_tan **+95.9%** (55.5 vs 28.3): tandem transfer nearly doubled
- p_oodc **+306.8%** (31.1 vs 7.6): OOD conditions 4x worse
- p_re **+291.3%** (24.7 vs 6.3): OOD Reynolds 4x worse
- p_in +4.7% (12.5 vs 12.0): in-distribution only slightly affected

**Why it failed so catastrophically:**

1. **Smoothness penalty conflicts with sharp gradients needed for generalization.** OOD conditions (different Re, camber, AoA) produce genuinely different pressure distributions with different suction peak heights and adverse pressure gradient profiles. The smoothness loss trains the model to suppress spatial pressure variation, which is exactly what's needed to capture OOD physics. Even with the LE/TE exclusion zone, the suction peak (typically at 5-15% chord) and trailing edge pressure recovery region have steep but physical gradients that the loss penalizes.

2. **Squared gradient penalty is scale-sensitive.** (dp/ds)² grows quadratically with pressure magnitude. At higher Re (OOD-Re) and stronger camber (OOD-cond), pressure gradients are inherently larger. The penalty disproportionately suppresses large-gradient predictions, which are exactly the OOD cases.

3. **The per-sample Python loop through all foils creates a gradient bottleneck.** With B batches × 2 foils × ~150 nodes sorted/indexed per call, this is ~10x slower than vectorized loss computation, and the sort operation introduces non-smooth gradients that may destabilize PCGrad.

4. **Redundancy with DCT frequency loss.** The DCT loss already penalizes high-frequency content (with proper frequency weighting). The Laplacian adds an aggressive spatial smoothness constraint on top, resulting in over-regularization.

5. **Same failure mode as arc-length surface loss (#2210).** Despite the LE/TE exclusion safeguard, the fundamental problem is identical: penalizing spatial pressure variation prevents the model from learning to represent the full range of pressure distributions.

### Suggested follow-ups

- **Do not pursue further smoothness regularization losses.** This is now the third experiment (arc-length #2210, DCT freq #2184, Laplacian #2232) testing surface pressure smoothness constraints. The pattern is clear: smoothness losses either help marginally (DCT, already merged at low weight) or catastrophically hurt (arc-length, Laplacian). The model's bottleneck is not noise/roughness in predictions.
- **If smoothness is still desired**, apply it as a very weak **eval-time post-processing** (Gaussian smooth on the surface ring) rather than as a training loss, to avoid corrupting the learned representation.
- Focus future experiments on **input feature engineering** (like wake deficit which worked) or **architectural improvements** rather than auxiliary loss terms.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2232 because ADVISOR: Closing — catastrophic OOD regression. p_oodc +307%, p_re +291%, p_tan +96%.

Your analysis is excellent and definitive. This is the third surface smoothness experiment (arc-length #2210, DCT freq #2184, Laplacian #2232). The pattern is now crystal clear:

**Surface smoothness constraints destroy OOD generalization.** The (dp/ds)² penalty disproportionately suppresses large-gradient predictions — which are EXACTLY the physically correct OOD pressure distributions (higher Re → steeper gradients, cambered foils → stronger suction peaks).

DCT freq loss barely helps at very low weight (0.05) because it's in frequency space. Spatial gradient penalties (arc-length, Laplacian) are catastrophic because they directly fight the physics of diverse pressure distributions.

Agreed: no more smoothness regularization losses. Moving you to a training schedule experiment..
